python/TestHarness/util.py

```diff
@@ -562,9 +622,138 @@ def didFail(self):
+        return status == self.bucket_skip
+
+    def isSilent(self):
+        """
+        Return boolean silent status
+        """
+        status = self.getStatus()
+        return status == self.bucket_silent
+
+    def isDeleted(self):
         """
-        Return boolean whether the test should be allowed to run or not
+        Return boolean deleted status
         """
         status = self.getStatus()
-        return not (status == self.bucket_deleted or status == self.bucket_skip or status == self.bucket_silent)
+        return status == self.bucket_deleted
+
+    def isFinished(self):
+        """
+        Return boolean finished status
+        """
+        status = self.getStatus()
+        return (status == self.bucket_finished or status is not self.bucket_pending)
+
+class StatusDependency():
+    """
+    A class used to determine if a tester can run based on other current tester statuses.
+
+    """
+
+    def __init__(self, tester, testers, options):
+        self.tester = tester
+        self.__testers = set(testers) - set([tester])
+        self.options = options
+
+    # A method that returns bool if the test can run or sets the appropriate status why it could not run
+    # or does nothing if the test needs to be placed back into the queue
+    def checkAndSetStatus(self):
+        if self.noRaceConditions():
+            if self.isRunnable():
+                return True
+            elif not self.isAppendable() and self.prereqsExists():
+                self.tester.setStatus('skipped dependency', self.tester.bucket_skip)
+            elif not self.prereqsExists():
+                self.tester.setStatus('unknown dependency', self.tester.bucket_fail)
+            else:
+                return
+        else:
+            self.tester.setStatus('OUTFILE RACE CONDITION', self.tester.bucket_fail)
+
+    # check if all prereq tests have passed and none were skipped/failed
+    def isRunnable(self):
+        if len(set(self.tester.getPrereqs()) - self._passing_tests()) == 0 or self.skipPrereqs():
+            return True
+
+    # check if any prereqs are not yet complete (means this test _will_ be runnable... just not right now)
+    def isAppendable(self):
+        if len(set(self.tester.getPrereqs()).intersection(self._skipped_tests())) == 0 or self.skipPrereqs():
+            return True
+
+    # check if all prereqs are available
+    def prereqsExists(self):
+        if len(set(self.tester.getPrereqs()) - set([x.getTestName() for x in self.__testers])) == 0:
+            return True
+
+    # return a set of passing tests
+    def _passing_tests(self):
+        passing = set([])
+        for tester in self.__testers:
+            if tester.didPass():
+                passing.add(tester.getTestName())
+        return passing
+
+    # return a set of finished non-passing or will be skipped tests
+    def _skipped_tests(self):
+        skipped_failed = set([])
+        for tmp_tester in self.__testers:
```

Comment:
self.__testers does not include the tester that entered this class... Then again, this class does not even get instanced if _this_ tester is skipped.

Hmmm.